producer.go

```diff
@@ -152,7 +151,51 @@ func (p *segmentProducer) configureSecurity(cfg Config) error {
 	return nil
 }
 
-func (p *segmentProducer) Send(key []byte, payload []byte, id []byte, ackFunc sdk.AckFunc) error {
+func (p *segmentProducer) Send(ctx context.Context, key []byte, payload []byte, id []byte, ackFunc sdk.AckFunc) error {
+	sendErr := p.sendRetryable(ctx, key, payload, id, ackFunc)
+	// If sendErr == nil, the message was successfully added to a batch,
+	// i.e. not actually sent.
+	// Because of that, we invoke ackFunc here only if sendErr != nil.
+	// NB: ackFunc will be invoked in a callback, when the batch is actually sent.
+	if sendErr == nil {
+		return nil
+	}
+
+	ackErr := ackFunc(sendErr)
```

Comment:
At this point `ackFunc` is still in the map `p.ackFuncs`, we should remove it.

Reply:
Good point!